### PR TITLE
Eng pr sr 14031 pointer type fix

### PR DESCRIPTION
**Pull request by bro4all**

### CODE REVIEW

**Comment by bro4all** [comment]
@swift-ci please test

**Comment by bro4all** [comment]
@hyp @zoecarver To review and merge this before I get added to the Apple org group

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

File: test/Interop/Cxx/templates/Inputs/function-templates.h [hunk 1]
```diff
@@ -61,6 +61,8 @@ template <class T> void constLvalueReference(const T &) {}
 
 template <class T> void forwardingReference(T &&) {}
 
+template <class T> void PointerTemplateParameter(T*){}
```
Comment (zoecarver) [inline test/Interop/Cxx/templates/Inputs/function-templates.h:64]: If you want you could add a few more tests, such as a method or a function template with multiple parameters.

But this is probably OK-- it's not a very big change.

**Review by zoecarver — APPROVED** [review]

**Comment by zoecarver** [comment]
@bro4all looks like there's a failing test: `Interop/Cxx/templates/defaulted-template-type-parameter-module-interface.swift`. It's failing because on line 20 we check that a pointer to a template param type is imported as an `OpaquePointer`. Should be pretty easy to fix, just update the test to match the new behavior.

**Comment by zoecarver** [comment]
Also, just a general note: it's a good idea to run all of the interop tests before making a patch. Just do `lit -sv test-macosx-arm64/Interop/Cxx/` from the swift build directory directory.

**Comment by bro4all** [comment]
@swift-ci please test

File: test/Interop/Cxx/templates/defaulted-template-type-parameter-module-interface.swift [hunk 2]
```diff
@@ -17,6 +17,6 @@
 // CHECK: func defaultedTemplateReferenceTypeParam<T>(_ t: UnsafeMutablePointer<T>)
 // The following types aren't imported correctly, but that does not have to do
 // with the fact that the template type paramaters are defaulted.
-// CHECK: func defaultedTemplatePointerTypeParam<T>(_ t: OpaquePointer!)
-// CHECK: func defaultedTemplatePointerReferenceTypeParam<T>(_ t: inout OpaquePointer!)
+// CHECK: func defaultedTemplatePointerTypeParam<T>(_ t: UnsafeMutablePointer!)
+// CHECK: func defaultedTemplatePointerReferenceTypeParam<T>(_ t: inout UnsafeMutablePointer!)
```
Comment (zoecarver) [inline test/Interop/Cxx/templates/defaulted-template-type-parameter-module-interface.swift:21]:
```suggestion
// CHECK: func defaultedTemplatePointerTypeParam<T>(_ t: UnsafeMutablePointer<T>!)
// CHECK: func defaultedTemplatePointerReferenceTypeParam<T>(_ t: inout UnsafeMutablePointer<T>!)
```

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please test Windows.

**Review by Kantitad — APPROVED** [review]

**Comment by zoecarver** [comment]
@swift-ci please smoke test and merge.

**Pull request merged**